File: bmp085/bmp085_debug.pl
#!/usr/bin/perl

# BOSCH BMP085 - Digital pressure sensor Library
# for Raspberry Pi
use strict;
use warnings;
use IO::File;
use Fcntl;
use Time::HiRes 'usleep';

# https://mirrors.edge.kernel.org/pub/linux/kernel/people/marcelo/linux-2.4/include/linux/i2c.h
use constant I2C_SLAVE       => 0x0703;
use constant I2C_SLAVE_FORCE => 0x0706;
use constant I2C_RDWR        => 0x0707;

use constant BMP085_I2C_ADDRESS => 0x77;

# OSS : standard = 1, high resolution = 2, ultra high resolution = 3
use constant BMP085_OVERSAMPLING_SETTING => 3;

# デバッグ時に1とすることで、計算途中の変数を端末に表示する
my $__debug = 1;

{
    my ( $t, $p ) = bmp085_getvalue();
    printf( "BMP085\ntemperature = %.1f deg-C, pressure = %.2f hPa\n",
            $t / 10, $p / 100 );
    exit;
}

# 温度・気圧を配列で返すsub
sub bmp085_getvalue {
    my ( $t, $p );

    eval {
        my $fh = IO::File->new( "/dev/i2c-1", O_RDWR );
        $fh->ioctl( I2C_SLAVE_FORCE, BMP085_I2C_ADDRESS );
        $fh->binmode();

        #
        # read calibration data
        #
        if ( $__debug == 1 ) { print "ac1 "; }
        my $ac1 = i2c_read_int( $fh, 0xaa );
        if ( $__debug == 1 ) { print "ac2 "; }
        my $ac2 = i2c_read_int( $fh, 0xac );
        if ( $__debug == 1 ) { print "ac3 "; }
        my $ac3 = i2c_read_int( $fh, 0xae );
        if ( $__debug == 1 ) { print "ac4 "; }
        my $ac4 = i2c_read_unsigned_int( $fh, 0xb0 );
        if ( $__debug == 1 ) { print "ac5 "; }
        my $ac5 = i2c_read_unsigned_int( $fh, 0xb2 );
        if ( $__debug == 1 ) { print "ac6 "; }
        my $ac6 = i2c_read_unsigned_int( $fh, 0xb4 );
        if ( $__debug == 1 ) { print "b1 "; }
        my $b1 = i2c_read_int( $fh, 0xb6 );
        if ( $__debug == 1 ) { print "b2 "; }
        my $b2 = i2c_read_int( $fh, 0xb8 );
        if ( $__debug == 1 ) { print "mb "; }
        my $mb = i2c_read_int( $fh, 0xba );
        if ( $__debug == 1 ) { print "mc "; }
        my $mc = i2c_read_int( $fh, 0xbc );
        if ( $__debug == 1 ) { print "md "; }
        my $md = i2c_read_int( $fh, 0xbe );

        #
        # read uncompensated temperature value
        #
        my @array_bytes = ( 0xF4, 0x2E );
        i2c_write_bytes( $fh, \@array_bytes );

        # wait 4.5 ms
        usleep(4500);
        if ( $__debug == 1 ) {
            print "write 0xF4, 0x2E, and wait 4.5msec\nut ";
        }
        my $ut = i2c_read_int( $fh, 0xf6 );

        #
        # read uncompensated pressure value
        #
        @array_bytes = ( 0xF4, 0x34 + ( BMP085_OVERSAMPLING_SETTING << 6 ) );
        i2c_write_bytes( $fh, \@array_bytes );

        # wait 7.5ms (if oss=1), 13.5ms (oss=2), 25.5ms (oss=3)
        usleep( 25.5 * 1000 );
        if ( $__debug == 1 ) {
            printf( "write 0xF4, 0x%02X, and wait 25.5msec\nup ",
                    0x34 + ( BMP085_OVERSAMPLING_SETTING << 6 ) );
        }
        my $up_bytes = i2c_read_bytes( $fh, 0xf6, 3 );
        my @up_array = unpack( "C*", $up_bytes );
        my $up = ( $up_array[0] << 16 | $up_array[1] << 8 | $up_array[2] )
          >> ( 8 - BMP085_OVERSAMPLING_SETTING );
        if ( $__debug == 1 ) { printf( "(long %ld)\n", $up ); }

        $fh->close();

        my $b5;
        ( $t, $b5 ) = bmp085_calc_temperature( $ut, $ac5, $ac6, $mc, $md );
        if ( $__debug == 1 ) { printf( "t = %f (deg-C)\n", $t / 10 ); }

        $p = bmp085_calc_pressure( $ac1, $ac2, $ac3, $ac4, $up, $b1, $b2, $b5 );
        if ( $__debug == 1 ) { printf( "p = %f (hPa)\n", $p / 100 ); }

    };
    if ($@) {
        die $@;
    }

    return ( $t, $p );

}

sub bmp085_calc_temperature {
    my ( $ut, $ac5, $ac6, $mc, $md ) = @_;
    my ( $t, $b5, $x1, $x2 );
    $x1 = int( ( ( $ut - $ac6 ) * $ac5 ) / ( 2**15 ) );
    $x2 = int( ( $mc * ( 2**11 ) ) / ( $x1 + $md ) );
    $b5 = $x1 + $x2;
    $t  = int( ( $b5 + 8 ) / ( 2**4 ) );
    if ( $__debug == 1 ) {
        print "x1 = $x1, x2 = $x2, b5 = $b5\n";
    }

    return ( $t, $b5 );
}

sub bmp085_calc_pressure {
    my ( $ac1, $ac2, $ac3, $ac4, $up, $b1, $b2, $b5 ) = @_;
    my ( $x1, $x2, $x3, $b3, $b4, $b6, $b7, $p );

    $b6 = int( $b5 - 4000 );

    $x1 = int( ( $b2 * ( $b6 * $b6 / ( 2**12 ) ) ) / ( 2**11 ) );
    $x2 = int( ( $ac2 * $b6 ) /                      ( 2**11 ) );
    $x3 = $x1 + $x2;
    $b3 =
      int( ( ( int( ($ac1) * 4 + $x3 ) << BMP085_OVERSAMPLING_SETTING ) + 2 ) /
           4 );
    if ( $__debug == 1 ) { print "x1=$x1, x2=$x2, x3=$x3, b3=$b3\n"; }

    $x1 = int( ( $ac3 * $b6 ) / ( 2**13 ) );
    $x2 = int( ( $b1 * ( $b6 * $b6 / ( 2**12 ) ) ) / ( 2**16 ) );
    $x3 = int( ( ( $x1 + $x2 ) + 2 ) / ( 2**2 ) );
    $b4 = int( $ac4 * int( $x3 + 32768 ) / ( 2**15 ) );
    if ( $__debug == 1 ) { print "x1=$x1, x2=$x2, x3=$x3, b4=$b4\n"; }

    $b7 = int( ( int($up) - $b3 ) * ( 50000 >> BMP085_OVERSAMPLING_SETTING ) );
    if   ( $b7 < 0x80000000 ) { $p = int( $b7 * 2 / $b4 ); }
    else                      { $p = int( $b7 / $b4 * 2 ); }
    if ( $__debug == 1 ) { print "b7=$b7, p=$p\n"; }

    $x1 = int( $p / ( 2**8 ) * $p / ( 2**8 ) );
    $x1 = int( ( $x1 * 3038 ) / ( 2**16 ) );
    $x2 = int( ( -7357 * $p ) / ( 2**16 ) );
    $p += int( ( $x1 + $x2 + 3791 ) / ( 2**4 ) );

    if ( $__debug == 1 ) { print "x1=$x1, x2=$x2, p=$p\n"; }

    return $p;

}

sub i2c_read_int {
    my ( $i2c, $bmp085_reg ) = @_;

    # アドレス 2Bytes をバイナリ形式の変数に格納
    my $buffer = pack( "C*", $bmp085_reg );

    # アドレス 2Bytes 送信
    #（bufferdのprintではなく、unbufferdのsyswrite利用）
    $i2c->syswrite($buffer);

    # データ 1Byte 受信
    #（bufferdのreadではなく、unbufferdのsysread利用）
    $i2c->sysread( $buffer, 2 );

    my $val0 = unpack( "n", $buffer ); # ビッグエンディアンのshort unsigned intとして解釈
    my $val = unpack( "s", pack( "S", $val0 ) ); # unsigned から signed に変換
    if ( $__debug == 1 ) {
        printf "(0x%02X) = 0x%02X, 0x%02X (int %d)\n", $bmp085_reg,
          unpack( "C", substr( $buffer, 0, 1 ) ),
          unpack( "C", substr( $buffer, 1, 1 ) ), $val;
    }

    return $val;
}

sub i2c_read_unsigned_int {
    my ( $i2c, $bmp085_reg ) = @_;

    # アドレス 2Bytes をバイナリ形式の変数に格納
    my $buffer = pack( "C*", $bmp085_reg );

    # アドレス 2Bytes 送信
    $i2c->syswrite($buffer);

    # データ 1Byte 受信
    $i2c->sysread( $buffer, 2 );

    my $val = unpack( "n", $buffer ); # ビッグエンディアンのshort unsigned intとして解釈
    if ( $__debug == 1 ) {
        printf "(0x%02X) = 0x%02X, 0x%02X (uint %u)\n", $bmp085_reg,
          unpack( "C", substr( $buffer, 0, 1 ) ),
          unpack( "C", substr( $buffer, 1, 1 ) ), $val;
    }

    return $val;
}

sub i2c_write_bytes {
    my ( $i2c, $ref_array, $count ) = @_;

    # アドレス 2Bytes をバイナリ形式の変数に格納
    my $buffer = pack( "C*", @{$ref_array} );

    # アドレス 2Bytes 送信
    $i2c->syswrite($buffer);

    return;
}

sub i2c_read_bytes {
    my ( $i2c, $bmp085_reg, $count ) = @_;

    # アドレス 2Bytes をバイナリ形式の変数に格納
    my $buffer = pack( "C*", $bmp085_reg );

    # アドレス 2Bytes 送信
    $i2c->syswrite($buffer);

    # データ 1Byte 受信
    if ( $i2c->sysread( $buffer, 3 ) != $count ) {
        print "(less data error) \n";
    }

    if ( $__debug == 1 ) {
        printf( "(0x%02X) = ", $bmp085_reg );
        for ( my $i = 0 ; $i < $count ; $i++ ) {
            printf( "0x%02X, ", unpack( "C", substr( $buffer, $i, 1 ) ) );
        }
    }
    return $buffer;
}

File: bmp085/bmp085_debug.c
/*
 * 
 * BOSCH BMP085 - Digital pressure sensor Library
 * for Raspberry Pi
 * 
 * (C)opyright INOUE Hirokazu
 * 2018 May 19
 * License: CC BY-SA v3.0 - http://creativecommons.org/licenses/by-sa/3.0/
 * 
 * compile : gcc bmp085_debug.c -lwiringPi
 * 
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <math.h>

//#define WIRING_PI

#ifdef WIRING_PI
#include <wiringPi.h>
#endif

#define BMP085_I2C_ADDRESS 0x77
// IMPORTANT ! : Rpi BCM_GPIO = 4  is equivalent to WiringPi GPIO = 7
#define GPIO_WRITE_PIN 7

const unsigned char BMP085_OVERSAMPLING_SETTING = 3;

// デバッグ時に1とすることで、計算途中の変数を端末に表示する
const int debug_info = 1;

/*
 * I2C デバイス EEPROM レジスタ i2c_reg よりint値を読み込む
 * (データ形式はビッグエンディアン）
 */
int i2c_read_int(int file, unsigned char i2c_reg)
{
    short int val = 0;

    char buf[1] = { i2c_reg };
    write(file, buf, 1);
    char data[2] = { 0, 0 };
    if(read(file, data, 2) != 2) {
        printf("error : less than 2 bytes data read from I2C device\n");
    }
    else {
        val = ((data[0] << 8) & 0xff00) | data[1];
        if(val & 0x8000) {
            val = -(val ^ 0xffff) + 1;
        }
    }
    if(debug_info) {
        printf("(0x%02X) = 0x%02X, 0x%02X (int %d)\n", i2c_reg, data[0], data[1], val);
    }
    return (val);
}

/*
 * I2C デバイス EEPROM レジスタ i2c_reg よりunsigned int値を読み込む
 * (データ形式はビッグエンディアン）
 */
int i2c_read_unsigned_int(int file, unsigned char i2c_reg)
{
    unsigned int val = 0;

    char buf[1] = { i2c_reg };
    write(file, buf, 1);
    char data[2] = { 0, 0 };
    if(read(file, data, 2) != 2) {
        printf("error : less than 2 bytes data read from I2C device\n");
    }
    else {
        val = ((data[0] << 8) & 0xff00) | data[1];
    }
    if(debug_info) {
        printf("(0x%02X) = 0x%02X, 0x%02X (uint %u)\n", i2c_reg, data[0], data[1], val);
    }
    return (val);
}

/*
 * I2C デバイスにバッファbufよりlengthバイト書き込む
 */
void i2c_write_bytes(int file, char *buf, int length)
{
    write(file, buf, length);
}

/*
 * I2C デバイスのEEPROMレジスタi2c_regより、バッファdataにlengthバイト読み込む
 */
int i2c_read_bytes(int file, unsigned char i2c_reg, char *data, int length)
{
    char buf[1] = { i2c_reg };
    write(file, buf, 1);

    if(read(file, data, length) != length) {
        printf("error : less than %d bytes data read from I2C device\n", length);
        return (0);
    }
    if(debug_info) {
        printf("(0x%02X) = ", i2c_reg);
        for (int i = 0; i < length; i++)
            printf("0x%02X, ", data[i]);
    }
    // 読み込んだデータBytes値を返す
    return (length);
}

int bmp085_calc_temperature(unsigned int ut, short unsigned int ac5, short unsigned int ac6,
                            short int mc, short int md, int *b5)
{
    int t, x1, x2;

    x1 = (int) (((ut - ac6) * ac5) / (1 << 15));
    x2 = (int) ((mc * (1 << 11)) / (x1 + md));
    *b5 = x1 + x2;
    t = (int) ((*b5 + 8) / (1 << 4));
    if(debug_info) {
        printf("x1 = %d, x2 = %d, b5 = %d\n", x1, x2, *b5);
    }

    return (t);
}

int bmp085_calc_pressure(short int ac1, short int ac2, short int ac3, short unsigned int ac4,
                         unsigned int up, short int b1, short int b2, int b5)
{
    int x1, x2, x3, b3;
    unsigned int b4;
    int b6;
    unsigned int b7;
    int p;

    b6 = (int) (b5 - 4000);

    x1 = (int) ((b2 * (b6 * b6 / (1 << 12))) / (1 << 11));
    x2 = (int) ((ac2 * b6) / (1 << 11));
    x3 = x1 + x2;
    b3 = (int) ((((int) ((ac1) * 4 + x3) << BMP085_OVERSAMPLING_SETTING) + 2) / 4);
    if(debug_info)
        printf("x1=%d, x2=%d, x3=%d, b3=%d\n", x1, x2, x3, b3);

    x1 = (int) ((ac3 * b6) / (1 << 13));
    x2 = (int) ((b1 * (b6 * b6 / (1 << 12))) / (1 << 16));
    x3 = (int) (((x1 + x2) + 2) / (1 << 2));
    b4 = (unsigned int) (ac4 * (int) (x3 + 32768) / (1 << 15));
    if(debug_info)
        printf("x1=%d, x2=%d, x3=%d, b4=%u\n", x1, x2, x3, b4);
    if(b4 == 0) {
        if(debug_info)
            printf("error : b4 = 0 (divide 0)\n");
        return (0);
    }

    b7 = (unsigned int) (((int) (up) - b3) * (50000 >> BMP085_OVERSAMPLING_SETTING));
    if(b7 < 0x80000000)
        p = (int) (b7 * 2 / b4);
    else
        p = (int) (b7 / b4 * 2);
    if(debug_info)
        printf("b7=%u, p=%d\n", b7, p);

    x1 = (int) (p / (1 << 8) * p / (1 << 8));
    x1 = (int) ((x1 * 3038) / (1 << 16));
    x2 = (int) ((-7357 * p) / (1 << 16));
    p += (int) ((x1 + x2 + 3791) / (1 << 4));

    if(debug_info)
        printf("x1=%d, x2=%d, p=%d\n", x1, x2, p);

    return (p);
}

void bmp085_getvalue(float *t, float *p)
{
    // GPIO 4 をBMP085の電源として利用する設定
#ifdef WIRING_PI
    if(wiringPiSetup() == -1) {
        printf("error wiringPi setup\n");
        return;
    }
    pinMode(GPIO_WRITE_PIN, OUTPUT);
    digitalWrite(GPIO_WRITE_PIN, HIGH);
#else
    system("gpio -g mode 4 out");
    system("gpio -g write 4 1");
#endif

    usleep(250 * 1000);         // BMP085の電源ON後0.25秒待つ

    // I2Cデバイスを開く
    int file;
    char *bus = "/dev/i2c-1";

    if((file = open(bus, O_RDWR)) < 0) {
        printf("Failed to open the bus. \n");
        exit(1);
    }
    ioctl(file, I2C_SLAVE, BMP085_I2C_ADDRESS);


    if(debug_info)
        printf("ac1 ");
    short int ac1 = i2c_read_int(file, 0xaa);

    if(debug_info)
        printf("ac2 ");
    short int ac2 = i2c_read_int(file, 0xac);

    if(debug_info)
        printf("ac3 ");
    short int ac3 = i2c_read_int(file, 0xae);

    if(debug_info)
        printf("ac4 ");
    short unsigned int ac4 = i2c_read_unsigned_int(file, 0xb0);

    if(debug_info)
        printf("ac5 ");
    short unsigned int ac5 = i2c_read_unsigned_int(file, 0xb2);

    if(debug_info)
        printf("ac6 ");
    short unsigned int ac6 = i2c_read_unsigned_int(file, 0xb4);

    if(debug_info)
        printf("b1 ");
    short int b1 = i2c_read_int(file, 0xb6);

    if(debug_info)
        printf("b2 ");
    short int b2 = i2c_read_int(file, 0xb8);

    if(debug_info)
        printf("mb ");
    short int mb = i2c_read_int(file, 0xba);

    if(debug_info)
        printf("mc ");
    short int mc = i2c_read_int(file, 0xbc);

    if(debug_info)
        printf("md ");
    short int md = i2c_read_int(file, 0xbe);

    char buf1[2] = { 0xF4, 0x2E };
    i2c_write_bytes(file, buf1, 2);

    usleep(4500);               // 4.5 msec
    if(debug_info) {
        printf("write 0xF4, 0x2E, and wait 4.5msec\nut ");
    }

    unsigned int ut = i2c_read_int(file, 0xf6);

    char buf2[2] = { 0xF4, 0x34 + (BMP085_OVERSAMPLING_SETTING << 6) };
    i2c_write_bytes(file, buf2, 2);

    // wait 7.5ms (if oss=1), 13.5ms (oss=2), 25.5ms (oss=3)
    usleep(25.5 * 1000);
    if(debug_info) {
        printf("write 0xF4, 0x%02X, and wait 25.5msec\nup ",
               0x34 + (BMP085_OVERSAMPLING_SETTING << 6));
    }
    char buf_up[3] = { };
    i2c_read_bytes(file, 0xf6, buf_up, 3);

    unsigned int up =
        (buf_up[0] << 16 | buf_up[1] << 8 | buf_up[2]) >> (8 - BMP085_OVERSAMPLING_SETTING);
    if(debug_info)
        printf("(long %d)\n", up);

    // i2cデバイスを閉じる
    close(file);
    // GPIO 4 をBMP085の電源として利用する設定
#ifdef WIRING_PI
    digitalWrite(GPIO_WRITE_PIN, LOW);
#else
    system("gpio -g write 4 0");
#endif

    int b5 = 0;

    *t = (float) bmp085_calc_temperature(ut, ac5, ac6, mc, md, &b5) / 10.0;

    if(debug_info)
        printf("t = %f (deg-C)\n", *t);

    *p = (float) bmp085_calc_pressure(ac1, ac2, ac3, ac4, up, b1, b2, b5) / 100.0;

    if(debug_info)
        printf("p = %f (hPa)\n", *p);

    return;
}

int main(int argc, char **argv)
{
    float t, p;

    bmp085_getvalue(&t, &p);
    printf("BMP085\ntemperature = %.1f deg-C, pressure = %.2f hPa\n", t, p);

    return 0;
}
